core/lib/spree/core/controller_helpers/payment_parameters.rb

```diff
@@ -0,0 +1,67 @@
+module Spree
+  module Core::ControllerHelpers::PaymentParameters
+    # This method handles the awkwardness of how the html forms are currently
+    # set up for frontend and admin.
+    #
+    # This method expects a params hash in the format of:
+    #
+    #  {
+    #    payment_source: {
+    #      # The keys here are spree_payment_method.id's
+    #      '1' => {...source attributes for payment method 1...},
+    #      '2' => {...source attributes for payment method 2...},
+    #    },
+    #    order: {
+    #      # Note that only a single entry is expected/handled in this array
+    #      payments_attributes: [
+    #        {
+    #          payment_method_id: '1',
+    #        },
+    #      ],
+    #      ...other params...
+    #    },
+    #    ...other params...
+    #  }
+    #
+    # And this method returns a new params hash in the format of:
+    #
+    #  {
+    #    order: {
+    #      payments_attributes: [
+    #        {
+    #          payment_method_id: '1',
+    #          source_attributes: {...source attributes for payment method 1...}
+    #        },
+    #      ],
+    #      ...other params...
+    #    },
+    #    ...other params...
+    #  }
+    #
+    def move_payment_source_into_payments_attributes(original_params)
+      params = original_params.deep_dup
+
+      # Step 1: Gather all the information and ensure all the pieces are there.
+
+      return params if params[:payment_source].blank?
+
+      payment_params = params[:order] &&
+        params[:order][:payments_attributes] &&
+        params[:order][:payments_attributes].first
```

Comment:
Yeah this is hidden, hacky and surprising.  But, I'm just maintaining the existing functionality here. (See [here](https://github.com/solidusio/solidus/pull/303/files#diff-2c3e70899d4f22d0569021b01b0e307fL354)). With the way Frontend's html form is set up there is not a good way to handle multiple payments (and their sources) and Frontend will only ever send a single payment at a time so this hack takes advantage of that. It would be really nice to see if we could rework Frontend's form but for the moment I'm just trying to avoid shaving that yak.